pkg/netflow/portrollup/portrollup.go

```diff
@@ -46,37 +59,34 @@ func NewEndpointPairPortRollupStore(portRollupThreshold int) *EndpointPairPortRo
 		// curStore and newStore map key is composed of `<SOURCE_IP>|<DESTINATION_IP>`
 		// SOURCE_IP and SOURCE_IP are converted from []byte to string.
 		// string is used as map key since we can't use []byte as map key.
-		curStore: make(map[string]*portRollupTracker),
-		newStore: make(map[string]*portRollupTracker),
+		curStore: make(map[string][]uint16),
+		newStore: make(map[string][]uint16),
 
 		portRollupThreshold: portRollupThreshold,
 	}
 }
 
-func (prs *EndpointPairPortRollupStore) getOrCreate(sourceAddr []byte, destAddr []byte) (*portRollupTracker, *portRollupTracker) {
-	key := buildStoreKey(sourceAddr, destAddr)
-	if _, ok := prs.curStore[key]; !ok {
-		prs.curStore[key] = newPortRollupTracker()
-	}
-	if _, ok := prs.newStore[key]; !ok {
-		prs.newStore[key] = newPortRollupTracker()
-	}
-	return prs.curStore[key], prs.newStore[key]
+// Add will record new sourcePort and destPort for a specific sourceAddr and destAddr
+func (prs *EndpointPairPortRollupStore) Add(sourceAddr []byte, destAddr []byte, sourcePort uint16, destPort uint16) {
+	prs.AddToStore(prs.curStore, sourceAddr, destAddr, sourcePort, destPort)
+	prs.AddToStore(prs.newStore, sourceAddr, destAddr, sourcePort, destPort)
 }
 
-func (prs *EndpointPairPortRollupStore) getPortTracker(sourceAddr []byte, destAddr []byte) *portRollupTracker {
-	return prs.curStore[buildStoreKey(sourceAddr, destAddr)]
-}
+// AddToStore will add ports to store
+func (prs *EndpointPairPortRollupStore) AddToStore(store map[string][]uint16, sourceAddr []byte, destAddr []byte, sourcePort uint16, destPort uint16) {
+	srcToDestKey := buildStoreKey(sourceAddr, destAddr, isSourceEndpoint, sourcePort)
+	destToSrcKey := buildStoreKey(sourceAddr, destAddr, isDestinationEndpoint, destPort)
 
-// Add will record new sourcePort and destPort for a specific sourceAddr and destAddr
-func (prs *EndpointPairPortRollupStore) Add(sourceAddr []byte, destAddr []byte, sourcePort uint16, destPort uint16) {
 	prs.mu.Lock()
-	defer prs.mu.Unlock()
-
-	// Double write to current port tracker and new port tracker
-	curTracker, newTracker := prs.getOrCreate(sourceAddr, destAddr)
-	curTracker.add(sourcePort, destPort, prs.portRollupThreshold)
-	newTracker.add(sourcePort, destPort, prs.portRollupThreshold)
+	sourceToDestPorts := len(store[srcToDestKey])
+	destToSourcePorts := len(store[destToSrcKey])
+	if sourceToDestPorts >= prs.portRollupThreshold || destToSourcePorts >= prs.portRollupThreshold {
+		prs.mu.Unlock()
+		return
+	}
+	store[srcToDestKey] = appendPort(store[srcToDestKey], destPort)
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏
Seems like we need to store twice now, once from src to dest, and once from dest to src. Won't this lead to more memory used, and would there be a way to limit this in the future?

Reply:
When new flows are being received, at the beginning we don't know if source or destination port is ephemeral, so, we need to track both.

But once we know that one of the two is ephemeral we can optimise by avoiding tracking the port that is ephemeral.
This optimisation will be done in the follow up PR here https://github.com/DataDog/datadog-agent/pull/13792